LiS/man/man4/bufmod.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: bufmod.4.man,v 0.9.2.3 2008-04-29 08:33:18 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:18 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: bufmod.4.man,v $
.\" Revision 0.9.2.3  2008-04-29 08:33:18  brian
.\" - update headers for Affero release
.\"
.\" Revision 0.9.2.2  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH BUFMOD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B bufmod
\- \fISTREAMS\fR null module
.\"
.\"
.SH SYNOPSIS
.PP
.HP 8
.BI "int " fd ;
.HP 8
.BI "int open(" \*(lq\fR/dev/nuls\fP\*(rq ", " \fR0\fP );
.HP 8
.BI "int ioctl(" fd ", " I_PUSH ", " \*(lqbufmod\*(rq );
.\"
.\"
.SH DESCRIPTION
.PP
.B bufmod
is a
.I STREAMS
module that is pushed over any
.IR Stream ,
.IR STREAMS -based
.BR pipe (4)
or
.IR STREAMS -based
.BR fifo (4).
.PP
Aside from being a
.I STREAMS
module,
.B bufmod
performs no actions other than to perform flow control and queueing of messages.
All messages passed from the
.I Stream head
toward the
.B bufmod
module are passed on downstream from the
.B bufmod
write-side
.B put
procedure (with
.BR putnext (9))
when flow control conditions are not in effect.  When flow controlled, messages are placed on the
write-side queue to be scheduled from the module's
.B service
procedure.
All messages received from the driver,
are passed on upstream from the
read-side
.B put
procedure, or scheduled for the read-side
.B service
procedure under flow control.
.\"
.\"
.SH USAGE
.PP
The primary purpose of
.B bufmod
is as a test module for use with the
.BR test-streams (8)
conformance test program,
and the
.BR perftest (8)
performance benchmarking program.  The other purpose of
.B bufmod
is as an example skeleton for a
.I STREAMS
pushable module.
.\"
.\"
.SH NOTICES
.PP
.BR bufmod
is implementation-specific to the
.B @PACKAGE_TITLE@
package.
.\"
.\"
.SH EXAMPLES
.PP
The source code for
.B bufmod
serves as an example skeleton for a
.I STREAMS
pushable module.
.\"
.\"
.SH MODULES
.PP
.B bufmod
is a
.I STREAMS
module named
.IR \*(lqbufmod\*(rq .
.\"
.\"
.SH "SEE ALSO"
.PP
.BR open (2s),
.BR ioctl (2),
.BR streamio (7),
.BR putnext (9).
.\"
.\"
.SH BUGS
.PP
.B bufmod
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B bufmod
is compatible with the null module described in the
.IR "SVR 4.0" ", \*(lq" "STREAMS Programmer's Guide" ",\*(rq"
.[
svr4spg
%P Section 6.0.9.1, "Null Module Example"
.]
with the following portability considerations:
.IP \(em \w'\(em\(em'u
The null module example of
.I "SVR 4.0"
does not queue messages and has no
.B service
procedure.  See also
.BR nullmod (4).
.IP \(em
.B bufmod
is specific to
.BR "@PACKAGE_TITLE@" ,
and will not be used by portable
.I STREAMS
modules, drivers and applications programs.
.IP \(em
The
.I SVR 4.0 SPG
.[
svr4spg
.]
does not name this module.  The name
.B bufmod
was chosen for
.BR "@PACKAGE_TITLE@" .
.IP \(em
.B bufmod
is used as an example skeleton module and for use by the
.BR test-streams (8)
conformance test program, and the
.BR perftest (8)
performance benchmarking program.  It will not be used by portable
.I STREAMS
modules, drivers or applications.
.PP
See
.BR STREAMS (9)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
.IR "SVR 4.0" .
.[
svr4spg
.]
Conformance is not verified.
.\"
.\"
.SH HISTORY
.PP
.B bufmod
first appeared in
.IR "SVR 4.0" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/bufmod.c
/*****************************************************************************

 @(#) $RCSfile: bufmod.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: bufmod.c,v $
 Revision 0.9.2.15  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 0.9.2.14  2007/08/15 04:57:58  brian
 - GPLv3 updates

 Revision 0.9.2.13  2007/08/14 10:46:56  brian
 - GPLv3 header update

 *****************************************************************************/

#ident "@(#) $RCSfile: bufmod.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] =
    "$RCSfile: bufmod.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $";

/*
 *  This is BUFMOD a STREAMS buffering module that performs no actions other than acting as a
 *  STREAMS module and buffering input and output.  Its purpose is primarily for testing and for
 *  serving as an example of the skeleton of a STREAMS module.
 *
 *  This is an absurdly simple module.
 *
 *  In addition, this module provide a specialized set of input-output controls for testing and
 *  verifying various internal STREAMS functions.
 */

#include <sys/LiS/module.h>	/* should be first */

#include <sys/LiS/config.h>

#include <sys/stream.h>
#include <sys/stropts.h>
#ifdef LINUX
#include <linux/errno.h>
#else
#include <sys/errno.h>
#endif
#include <sys/cmn_err.h>
#include <sys/osif.h>

#define BUFMOD_DESCRIP		"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define BUFMOD_COPYRIGHT	"Copyright (c) 1997-2008 OpenSS7 Corporation.  All Rights Reserved."
#define BUFMOD_REVISION		"LfS $RCSfile: bufmod.c,v $ $Name:  $($Revision: 0.9.2.15 $) $Date: 2008-04-29 08:33:11 $"
#define BUFMOD_DEVICE		"SVR 4.2 Buffer Module (BUFMOD) for STREAMS"
#define BUFMOD_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define BUFMOD_LICENSE		"GPL"
#define BUFMOD_BANNER		BUFMOD_DESCRIP		"\n" \
				BUFMOD_COPYRIGHT	"\n" \
				BUFMOD_REVISION		"\n" \
				BUFMOD_DEVICE		"\n" \
				BUFMOD_CONTACT		"\n"
#define BUFMOD_SPLASH		BUFMOD_DEVICE		" - " \
				BUFMOD_REVISION		"\n"

#if defined LIS && defined MODULE
#define CONFIG_STREAMS_BUFMOD_MODULE MODULE
#endif

#ifdef CONFIG_STREAMS_BUFMOD_MODULE
MODULE_AUTHOR(BUFMOD_CONTACT);
MODULE_DESCRIPTION(BUFMOD_DESCRIP);
MODULE_SUPPORTED_DEVICE(BUFMOD_DEVICE);
MODULE_LICENSE(BUFMOD_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-bufmod");
#endif
#endif

#ifndef CONFIG_STREAMS_BUFMOD_NAME
#define CONFIG_STREAMS_BUFMOD_NAME BUFMOD__MOD_NAME
//#error "CONFIG_STREAMS_BUFMOD_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_BUFMOD_MODID
#define CONFIG_STREAMS_BUFMOD_MODID BUFMOD__ID
//#error "CONFIG_STREAMS_BUFMOD_MODID must be defined."
#endif

modID_t modid = CONFIG_STREAMS_BUFMOD_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for BUFMOD.");

#ifdef MODULE_ALIAS
#ifdef LFS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_BUFMOD_MODID));
MODULE_ALIAS("streams-module-bufmod");
#endif
#endif

STATIC struct module_info bufmod_minfo = {
	.mi_idnum = CONFIG_STREAMS_BUFMOD_MODID,
	.mi_idname = CONFIG_STREAMS_BUFMOD_NAME,
	.mi_minpsz = 0,
	.mi_maxpsz = 4096,
	.mi_hiwat = 5120,
	.mi_lowat = 1024,
};

STATIC struct module_stat bufmod_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
STATIC struct module_stat bufmod_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

#ifndef unlikely
#define unlikely(__x) (__x)
#endif
#ifndef likely
#define likely(__x) (__x)
#endif

/* 
 *  -------------------------------------------------------------------------
 *
 *  PUT routines
 *
 *  -------------------------------------------------------------------------
 */

STATIC int _RP
bufmod_wput(queue_t *q, mblk_t *mp)
{
	if (unlikely(mp->b_datap->db_type == M_FLUSH)) {
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
		}
	}
	if (likely(mp->b_datap->db_type >= QPCTL || (q->q_first == NULL && !(q->q_flag & QRUNNING)
						     && bcanputnext(q, mp->b_band)))) {
		putnext(q, mp);
		return (0);
	}
	/* always buffer, always schedule out of service procedure for testing */
	if (unlikely(putq(q, mp) == 0)) {
		mp->b_band = 0;
		putq(q, mp);	/* this must succeed */
	}
	return (0);
}

STATIC int _RP
bufmod_rput(queue_t *q, mblk_t *mp)
{
	if (unlikely(mp->b_datap->db_type == M_FLUSH)) {
		if (mp->b_rptr[0] & FLUSHR) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
		}
	}
	if (likely(mp->b_datap->db_type >= QPCTL || (q->q_first == NULL && !(q->q_flag & QRUNNING)
						     && bcanputnext(q, mp->b_band)))) {
		putnext(q, mp);
		return (0);
	}
	if (unlikely(putq(q, mp) == 0)) {
		mp->b_band = 0;
		putq(q, mp);	/* this must succeed */
	}
	return (0);
}

STATIC int _RP
bufmod_srv(queue_t *q)
{
	mblk_t *mp;

	while ((mp = getq(q))) {
		if (likely(mp->b_datap->db_type >= QPCTL || bcanputnext(q, mp->b_band))) {
			putnext(q, mp);
			continue;
		}
		putbq(q, mp);
		break;
	}
	return (0);
}

/*
 *  -------------------------------------------------------------------------
 *
 *  OPEN and CLOSE
 *
 *  -------------------------------------------------------------------------
 */
STATIC int _RP
bufmod_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *crp)
{
	queue_t *wq;

	if (q->q_ptr != NULL) {
		return (0);	/* already open */
	}
	wq = WR(q);
	if (sflag == MODOPEN || wq->q_next != NULL) {
		/* inherit packet sizes of what we are pushed over */
		wq->q_minpsz = wq->q_next->q_minpsz;
		wq->q_maxpsz = wq->q_next->q_maxpsz;
		q->q_ptr = wq->q_ptr = q;	/* just set it to something */
		qprocson(q);
		return (0);
	}
	return (EIO);		/* can't be opened as driver */
}
STATIC int _RP
bufmod_close(queue_t *q, int oflag, cred_t *crp)
{
	(void) oflag;
	(void) crp;
	if (!q->q_ptr)
		return (ENXIO);
	qprocsoff(q);
	q->q_ptr = WR(q)->q_ptr = NULL;
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  Registration and initialization
 *
 *  -------------------------------------------------------------------------
 */
STATIC struct qinit bufmod_rinit = {
	.qi_putp = bufmod_rput,
	.qi_srvp = bufmod_srv,
	.qi_qopen = bufmod_open,
	.qi_qclose = bufmod_close,
	.qi_minfo = &bufmod_minfo,
	.qi_mstat = &bufmod_rstat,
};

STATIC struct qinit bufmod_winit = {
	.qi_putp = bufmod_wput,
	.qi_srvp = bufmod_srv,
	.qi_minfo = &bufmod_minfo,
	.qi_mstat = &bufmod_wstat,
};

STATIC struct streamtab bufmod_info = {
	.st_rdinit = &bufmod_rinit,
	.st_wrinit = &bufmod_winit,
};

#ifdef MODULE
STATIC
#endif
int __init
bufmod_init(void)
{
	int err;

#ifdef MODULE
	printk(KERN_INFO BUFMOD_BANNER);
#else
	printk(KERN_INFO BUFMOD_SPLASH);
#endif
	bufmod_minfo.mi_idnum = modid;
	if ((err = lis_register_strmod(&bufmod_info, CONFIG_STREAMS_BUFMOD_NAME)) < 0)
		return (err);
	if (modid == 0 && err > 0)
		modid = err;
	if ((err = lis_register_module_qlock_option(modid, LIS_QLOCK_NONE)) < 0) {
		lis_unregister_strmod(&bufmod_info);
		return (err);
	}
	return (0);
}

#ifdef MODULE
STATIC
#endif
void __exit
bufmod_exit(void)
{
	int err;

	if ((err = lis_unregister_strmod(&bufmod_info)) < 0)
		return (void) (err);
	return (void) (0);
}

#ifdef MODULE
module_init(bufmod_init);
module_exit(bufmod_exit);
#endif
